File: src/client/client_accept_manager.rs
use std::collections::HashMap;
use std::sync::Arc;
use lazy_static::lazy_static;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::{TcpListener, TcpStream};
use tokio::sync::Mutex;
use crate::client::header_util;

lazy_static! {
    pub static ref CLIENT_SOCKET_MAP: Arc<Mutex<HashMap<String, TcpStream>>> = Arc::new(Mutex::new(HashMap::new()));
}

///开始
pub async fn start() {
    tokio::spawn(async {
        accept()
    });
}

///监听客户端连接
async fn accept() {

    // 绑定到指定地址和端口
    let bind_result = TcpListener::bind("0.0.0.0:1681").await;
    if let Err(e) = bind_result {
        eprintln!("创建TcpListener失败: {}", e);
        return;
    }

    //等待客户端连接
    println!("-->监听客户端连接开始");
    let listener = bind_result.unwrap();
    loop {
        let accept_result = listener.accept().await;
        if let Err(e) = accept_result {
            eprintln!("监听客户端监听失败: {}", e);

            //tokio::time::sleep(tokio::time::Duration::from_secs_f32(10.0)).await;
            return;
        }
        println!("-->接收到客户端连接请求");
        let (tcp_stream, _) = accept_result.unwrap();

        tokio::spawn(async {
            handle(tcp_stream)
        });
    }

    // println!("New connection: {:?}", socket.peer_addr());



    // let mut global_stream = CLIENT_SOCKET_MAP.lock().await;
    // global_stream.insert(String::from("TT"), socket);


    println!("-->监听客户端连接结束")
    // //开启异步任务
    // tokio::spawn(async {
    //     let mut global_stream = PIPELINE_SOCKET.lock().await;
    //
    //     if let Some(ref mut stream) = *global_stream {
    //         handle(stream).await;
    //     } else {
    //         println!("TcpStream is not set");
    //     }
    // });
    // }
}


/**
 * 分配连接
 * @param socketClient 与客户端的连接
 */
async fn handle(mut tcp_stream: TcpStream){

//保持长连接
// socketClient.keepAlive = true

//得到输入流
// val clientIStream = socketClient.inputStream
//
// //读取连接的第一个数据,设置超时,避免恶意连接
// socketClient.soTimeout = 5000

//println("-->读取标记开始${System.currentTimeMillis()}")

    // 创建一个长度为 1 的缓冲区
    let mut buf = [0; 1];

    //读取第一个标记字节,通过该自己判断该连接类型
    tcp_stream.read(&mut buf).await;

    let flag = buf[0];
    match flag{

        //标记该连接为:服务器端往客户端发送指令的连接
        header_util::CLIENT_TO_SERVER_MAIN_CONNECTION => {

        }

        //创建客户端Socket连接池
        2 => {
            // TCPPoolManager.add(socketClient)
        }
        _ => {
            tcp_stream.shutdown().await;
        }
    }
}

// ///等待客户端连接
// async fn handle(socket: &mut TcpStream) {
//     loop {
//
//         // // 创建一个缓冲区来读取数据
//         // let mut buffer = [0; 8 * 1024];
//         //
//         // // 从客户端读取数据
//         // let len_result = socket.read(&mut buffer).await;
//         // if let Err(e) = len_result {
//         //     eprintln!("读取数据失败: {}", e);
//         //     break;
//         // }
//         // let len = len_result.unwrap();
//         // if len == 0{//客户端已经关闭了
//         //     break;
//         // }
//         //
//         // // 打印接收到的数据
//         // println!("Received: {}", String::from_utf8_lossy(&buffer[..len]));
//
//         tokio::time::sleep(tokio::time::Duration::from_secs_f32(2.0)).await;
//
//         // 回应客户端
//         let write_result = socket.write(b"Hello from server!").await;
//         if let Err(e) = write_result {
//             eprintln!("写入数据失败: {}", e);
//             break;
//         }
//     }
// }

File: client/ClientAcceptManager/ClientAcceptManager.go
package ClientAcceptManager

import (
	"DairoNPS/client/ClientSessionManager"
	"DairoNPS/client/HeaderUtil"
	"DairoNPS/constant/CLSConfig"
	"DairoNPS/pool/TCPPoolManager"
	"fmt"
	"log"
	"net"
	"time"
)

// 接受并HOLD住客户端的socket

/**
 * Socket服务端
 */
var socketServer net.Listener

// Start 启动
func Start() {
	//go accept()
	//time.Sleep(5 * time.Second)
	accept()
}

/**
 * 监听客户端连接
 */
func accept() {
	listen, err := net.Listen("tcp", fmt.Sprintf(":%d", CLSConfig.CLIENT_TO_SERVER_TCP_PORT))
	socketServer = listen

	//最终关闭监听
	defer func(listen net.Listener) {
		err := listen.Close()
		if err != nil {
			log.Fatalf("%q\n", err)
		}
	}(listen)

	//启动失败
	if err != nil {
		println("启动失败，请参考错误信息。")
		log.Fatalf("%q\n", err)
	}

	fmt.Printf("端口:%d监听成功。\n", CLSConfig.CLIENT_TO_SERVER_TCP_PORT)
	//go func() {
	//	time.Sleep(3 * time.Second)
	//	listen.Close()
	//}()
	for {

		//等待客户端连接
		fmt.Println("-->监听客户端连接")
		socketClient, err := socketServer.Accept()
		if err != nil {
			log.Fatalf("%q\n", err)
		}
		fmt.Println("-->接收到客户端连接请求")
		go handle(socketClient)
	}
}

/**
 * 分配连接
 * @param socketClient 与客户端的连接
 */
func handle(socketClient net.Conn) {

	//保持长连接
	//socketClient.keepAlive = true

	//得到输入流
	//val clientIStream = socketClient.inputStream

	//读取连接的第一个数据,设置超时,避免恶意连接
	socketClient.SetReadDeadline(time.Now().Add(300 * time.Second))
	//socketClient.SetReadDeadline(time.Time{})

	//println("-->读取标记开始${System.currentTimeMillis()}")
	//读取第一个标记字节,通过该自己判断该连接类型

	flagData := make([]byte, 1)
	len, err := socketClient.Read(flagData)
	if len == 0 || err != nil {
		log.Println("-->从客户端读取数据超时")

		//没必要继续执行，直接关闭客户端连接
		socketClient.Close()
		return
	}
	//println("-->读取标记结束${System.currentTimeMillis()}:$flag")
	switch flagData[0] {

	//标记该连接为:服务器端往客户端发送指令的连接
	case HeaderUtil.CLIENT_TO_SERVER_MAIN_CONNECTION:
		ClientSessionManager.Validate(socketClient)

	//创建客户端Socket连接池
	case HeaderUtil.SERVER_TCP_POOL_REQUEST:
		TCPPoolManager.Add(socketClient)
	}
}
